### Imports

In [2]:
import json
import numpy as np
import faiss
import requests
from sentence_transformers import CrossEncoder

from sentence_transformers import SentenceTransformer

### Load FAISS + metadata

In [4]:
INDEX_PATH = "../data/embeddings/faiss.index"
META_PATH = "../data/embeddings/metadata.json"

index = faiss.read_index(INDEX_PATH)

with open(META_PATH, "r", encoding="utf-8") as f:
    chunks = json.load(f)

print(f"Loaded {len(chunks)} chunks")

Loaded 8361 chunks


### Load embedding model

In [6]:
model = SentenceTransformer("BAAI/bge-small-en-v1.5")

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

### Retrieval

In [8]:
def search(query, k=20):
    q_emb = model.encode([query]).astype("float32")
    distances, indices = index.search(q_emb, k)
    return [chunks[i] for i in indices[0]]

### Deduplication

In [10]:
def unique_results(results, max_results=5):
    seen = set()
    unique = []

    for r in results:
        text = r["text"]

        if text[:200] not in seen:
            unique.append(r)
            seen.add(text[:200])

        if len(unique) >= max_results:
            break

    return unique

### Build context

In [12]:
def build_context(results):
    return "\n\n".join([r["text"] for r in results])

### Prompt (VERY IMPORTANT for local models)

In [14]:
def build_prompt(query, context):
    return f"""<|begin_of_text|><|start_header_id|>system<|end_header_id|>
You are an ML project expert answering questions about a technical portfolio.

CRITICAL RULES:
1. Answer ONLY using information from the context below
2. If the context contains a specific answer — give it directly and concisely
3. If multiple options are mentioned but one is identified as "best" or "final" — state that one
4. Use exact numbers, metrics, and model names from the context
5. Keep answers to 1-3 sentences maximum
6. If the answer is NOT in the context, say exactly: "I don't know"
7. Never say "the documentation does not specify" — either give the answer or say "I don't know"

<|eot_id|><|start_header_id|>user<|end_header_id|>
CONTEXT:
---
{context}
---

QUESTION: {query}

ANSWER (concise, factual, 1-3 sentences):
<|eot_id|><|start_header_id|>assistant<|end_header_id|>"""

### Call local LLM (Ollama)

In [16]:
def ask_llm(prompt):
    response = requests.post(
        "http://localhost:11434/api/generate",
        json={
            "model": "phi3:mini",
            "prompt": prompt,
            "stream": False,
            "options": {
                "num_predict": 512,
                "temperature": 0.1
            }
        }
    )
    return response.json()["response"]

### Full RAG pipeline

In [18]:
def rag(query):
    # 1. Retrieve
    results = search(query, k=20)
    
    # 2. Deduplicate
    results = unique_results(results, max_results=5)
    
    # 3. Build context
    context = build_context(results)
    
    # 4. Prompt
    prompt = build_prompt(query, context)
    
    # 5. LLM
    answer = ask_llm(prompt)
    
    return answer, results

### Test

In [20]:
query = "What model is used in churn prediction in api?"

answer, results = rag(query)

print("ANSWER:\n")
print(answer)

ANSWER:

The project uses CatBoost for churn prediction. The model aims to achieve recall greater than 85% on test data and is saved using cloudpickle for production use, with the service containerized in Docker. Additionally, FastAPI endpoints are created for predictions, including API key authentication with role-based access control (user, admin, dashboard) and rate limiting set between 100 to 1000 requests per minute.


In [21]:
query = "What is the recall of the best churn prediction model?"

answer, results = rag(query)

print("ANSWER:\n")
print(answer)

ANSWER:

The CatBoost model achieved the highest recall of 96% for churn prediction. This indicates that it successfully identified most customers who were likely to leave, with only a 4% miss rate. The high recall is critical in this context as missing a customer's potential departure has significant costs associated with lost revenue and negative word-of-mouth impact.


In [22]:
# 3. Traffic system architecture
query = "Describe the architecture of the traffic sign detection system."
answer, results = rag(query)
print("\n--- Q3 ---")
print("ANSWER:\n", answer)


--- Q3 ---
ANSWER:
 The Traffic Sign Detection System uses a two-stage hybrid architecture combining YOLO for initial detection and ResNet-18 models specialized in classifying the detected signs into one of 55 classes. The system achieves high accuracy with mAP50 at 94.07% and Recall at 91.71%.


In [23]:
# 4. Improvement vs baseline
query = "How is the traffic sign detection system better than a baseline YOLO model?"
answer, results = rag(query)
print("\n--- Q4 ---")
print("ANSWER:\n", answer)


--- Q4 ---
ANSWER:
 The traffic sign detection system outperforms the baseline YOLO model by achieving a **94.07% mAP50** and **91.71% Recall**, indicating higher accuracy in detecting signs compared to the strong performance (mAP50 = 0s of 81.3%) achieved with just baseline YOLOv11n for traffic sign detection tasks.


In [24]:
# 5. Why resnet
query = "Why resnet?"
answer, results = rag(query)
print("\n--- Q5 ---")
print("ANSWER:\n", answer)


--- Q5 ---
ANSWER:
 ResNet was chosen for its proven effectiveness in fine-grained classification tasks and because it has been pretrained on a large dataset, which can help with better feature extraction from the images. Additionally, ResNet's architecture allows easy modification of layers to suit specific needs like adding custom fully connected (FC) layers tailored for 55 signs total in this project.


In [25]:
# 6. Full pipeline
query = "Describe the full pipeline of the churn prediction system."
answer, results = rag(query)
print("\n--- Q6 ---")
print("ANSWER:\n", answer)


--- Q6 ---
ANSWER:
 The churn prediction pipeline consists of six automated steps: TotalChargesCleaner for data cleaning; FeatureEngineer to extract features from the raw customer data. CategoricalEncoder is used for encoding categorical variables, while DropRedundant removes redundant information and NumericalScaler standardizes numerical values. The final output of this pipeline provides churn predictions in a single call when loaded into memory at startup using .pkl format.


In [26]:
# 7. Backend tech
query = "What technologies were used to build the churn prediction API?"
answer, results = rag(query)
print("\n--- Q7 ---")
print("ANSWER:\n", answer)


--- Q7 ---
ANSWER:
 The churn prediction service was built using FastAPI for the backend API development and Docker containerization. The machine learning model employed is CatBoost, which achieved a recall of over 85% on test data. Additionally, Streamlit was used to create an interactive dashboard for visualizing predictions.


In [27]:
# 7. Backend tech
query = "What technologies were used to build the sign detection API?"
answer, results = rag(query)
print("\n--- Q8 ---")
print("ANSWER:\n", answer)


--- Q8 ---
ANSWER:
 The traffic sign detection service was built using FastAPI and Docker. It features a YOLO + ResNet cascade architecture for detecting 55 classes of traffic signs with high accuracy in real-time, utilizing WebSocket technology to communicate between the client (browser) and server.


In [28]:
# 8. Skills
query = "What machine learning skills does this portfolio demonstrate?"
answer, results = rag(query)
print("\n--- Q9 ---")
print("ANSWER:\n", answer)


--- Q9 ---
ANSWER:
 The portfolio demonstrates skills in predictive modeling using supervised machine learning to assess the engagement level of educational content, as well as experience with Python-based data science workflows. It also shows proficiency in handling real-world analytical problems and creating reproducible results for applied projects within a Data Science & Machine Learning context.


In [29]:
# 9. Deployment
query = "How is the churn project deployed?"
answer, results = rag(query)
print("\n--- Q10 ---")
print("ANSWER:\n", answer)


--- Q10 ---
ANSWER:
 The deployment of the churn prediction service is done using Docker Compose on an Ubuntu VPS. It consists of two containers: one for FastAPI running at port 8000 and another for Streamlit dashboard, which operates through a reverse proxy listening to requests from clients via HTTPS (port 443).


In [30]:
# 10. Production readiness
query = "What makes this portfolio production-ready?"
answer, results = rag(query)
print("\n--- Q11 ---")
print("ANSWER:\n", answer)


--- Q11 ---
ANSWER:
 This ML project is considered production-ready due to its comprehensive features: an interactive Streamlit dashboard for real-time monitoring, a robust labeling system with SQLite persistence and batch operations, a complete machine learning pipeline using custom scikit-learn transformers and CatBoost models that processes raw data into predictions in one call. Additionally, the project is containerized with Docker Compose and deployed publicly while ensuring zero downtime during model updates.


In [31]:
# 11. Why recall
query = "Why was recall chosen as the main metric for churn prediction?"
answer, results = rag(query)
print("\n--- Q12 ---")
print("ANSWER:\n", answer)


--- Q12 ---
ANSWER:
 Recall was chosen as the main metric for churn prediction because in this context it is critical to correctly identify most customers who are likely to leave, and a high recall rate minimizes false negatives. The final CatBoost model achieved an impressive 96% recall, indicating that only 4% of potential churners were missed by the system.


In [32]:
# 12. Summary
query = "Summarize the churn project in 2 sentences."
answer, results = rag(query)
print("\n--- Q13 ---")
print("ANSWER:\n", answer)


--- Q13 ---
ANSWER:
 The Telco Customer Churn project aimed to predict customer churn and provide retention strategies using machine learning models, focusing on recall as a priority metric due to class imbalance. The team performed data splitting with an 80/20 train-test split while handling missing values by setting `TotalCharges` to zero when tenure is zero, ensuring no customers without billing history were excluded from the analysis.


In [33]:
# 13. 
query = "What is the best algorithm for the churn problem??"
answer, results = rag(query)
print("\n--- Q14 ---")
print("ANSWER:\n", answer)


--- Q14 ---
ANSWER:
 Boosting models such as XGBoost, LightGBM, and CatBoost achieved the highest recall (>0.88), making them the best performers for churn prediction in this context. They excelled at detecting most churn cases among all tested algorithms.


In [34]:
query = "Tell me about portfolio in general."
answer, results = rag(query)
print("\n--- Q14 ---")
print("ANSWER:\n", answer)


--- Q14 ---
ANSWER:
 The Data Science & Machine Learning Portfolio is a comprehensive collection of projects that demonstrate my progression from understanding data science fundamentals to applying advanced analytical and machine learning solutions in real-world scenarios. It emphasizes practical problem-solving, clean workflows, and reproducibility using Python for data analysis.


In [35]:
query = "What's the documentation about? Tell me how the Churn prediction service works?"
answer, results = rag(query)
print("\n--- Q15 ---")
print("ANSWER:\n", answer)


--- Q15 ---
ANSWER:
 The project is a production-ready machine learning service designed to predict customer churn for telecommunications companies. It uses FastAPI, CatBoost with a recall of 95.7%, and provides real-time predictions through an API as well as data visualization via Streamlit dashboard. The system also includes manual labeling for ground truth collection and automated retraining to continuously improve the model's accuracy.


In [62]:
query = "tell me about how api for churn work?"
answer, results = rag(query)
print("\n--- Q15 ---")
print("ANSWER:\n", answer)


--- Q15 ---
ANSWER:
 The Churn Prediction API allows for single and batch predictions of customer churn using a FastAPI endpoint. It requires authentication, supports role-based access with an admin or user role, includes rate limiting to prevent abuse (100-1000 requests per minute), and provides Swagger documentation for ease of use. The API returns binary prediction results along with probability scores indicating the likelihood of churn based on customer features.


## Summary 

This notebook implements a **complete RAG (Retrieval-Augmented Generation) pipeline** for querying an ML portfolio.

---

###  Main Components

| Component | What it does |
|-----------|---------------|
| **FAISS + metadata** | Loads pre-computed embeddings and 8,361 text chunks |
| **BGE embedding model** | Encodes user queries into vectors (`BAAI/bge-small-en-v1.5`) |
| **Semantic search** | Finds top-20 relevant chunks using cosine similarity |
| **Deduplication** | Removes duplicate chunks (keeps top-5 unique) |
| **LLM (phi3:mini)** | Generates concise answers via local Ollama |
| **Strict prompt** | Forces factual answers from context only |

---

###  Pipeline Flow
Query → Embed → FAISS Search (k=20) → Deduplicate (top-5) → Build Context → LLM (phi3:mini) → Answer


---

###  Key Functions

| Function | Purpose |
|----------|---------|
| `search(query, k=20)` | Retrieves chunks from FAISS index |
| `unique_results(results, max_results=5)` | Removes near-duplicate chunks |
| `build_context(results)` | Joins chunks into a single context string |
| `build_prompt(query, context)` | Creates a strict, fact-based prompt |
| `ask_llm(prompt)` | Calls local Ollama with phi3:mini |
| `rag(query)` | Orchestrates the full pipeline |

---

###  Prompt Rules (Critical)

1. Answer ONLY using context
2. Give direct, concise answers
3. If "best" or "final" is mentioned — state that one
4. Use exact numbers and model names
5. Keep answers to 1-3 sentences
6. If not in context → "I don't know"
7. Never say "the documentation does not specify"

---

###  Summary

This notebook turns the portfolio into a **question-answering system**. User asks a question → system finds relevant chunks → LLM answers based ONLY on those chunks. Results are factual, concise, and grounded in the actual portfolio content.